src/services/codeFixProvider.ts

```diff
@@ -1,40 +1,56 @@
 /* @internal */
 namespace ts {
-    export interface CodeFix {
+    export interface CodeFixRegistration {
         errorCodes: number[];
-        getCodeActions(context: CodeFixContext): CodeAction[] | undefined;
+        getCodeActions(context: CodeFixContext): CodeFix[] | undefined;
+        actionIds?: string[];
+        getAllCodeActions?(context: CodeFixAllContext): CodeActionAll;
     }
 
-    export interface CodeFixContext extends textChanges.TextChangesContext {
-        errorCode: number;
+    export interface CodeFixContextBase extends textChanges.TextChangesContext {
         sourceFile: SourceFile;
-        span: TextSpan;
         program: Program;
         host: LanguageServiceHost;
         cancellationToken: CancellationToken;
     }
 
+    export interface CodeFixAllContext extends CodeFixContextBase {
+        actionId: {};
+    }
+
+    export interface CodeFixContext extends CodeFixContextBase {
+        errorCode: number;
+        span: TextSpan;
+    }
+
     export namespace codefix {
-        const codeFixes: CodeFix[][] = [];
+        const codeFixes: CodeFixRegistration[][] = [];
+        const groups = createMap<CodeFixRegistration>();
 
-        export function registerCodeFix(codeFix: CodeFix) {
-            forEach(codeFix.errorCodes, error => {
+        export function registerCodeFix(codeFix: CodeFixRegistration) {

```

Comment:
Now that there's a `CodeFix` type, `codeFix` is a confusing name for a `CodeFixRegistration`.